eclair-core/src/main/scala/fr/acinq/eclair/payment/send/Recipient.scala

```diff
@@ -0,0 +1,227 @@
+        case None => NodePayload(nodeId, FinalPayload.Standard.createPayload(route.amount, totalAmount, expiry, paymentSecret, paymentMetadata_opt, customTlvs))
+      }
+      OutgoingPaymentPacket.buildPayloads(route.amount, expiry, Seq(finalPayload), hops)
+    })
+  }
+}
+
+object ClearRecipient {
+  def apply(invoice: Bolt11Invoice, totalAmount: MilliSatoshi, expiry: CltvExpiry, customTlvs: Seq[GenericTlv]): ClearRecipient = {
+    ClearRecipient(invoice.nodeId, invoice.features, totalAmount, expiry, invoice.paymentSecret, invoice.extraEdges, invoice.paymentMetadata, None, customTlvs)
+  }
+}
+
+/** A payment recipient that doesn't expect to receive a payment and can directly be found in the routing graph. */
+case class SpontaneousRecipient(nodeId: PublicKey,
+                                amount: MilliSatoshi,
+                                expiry: CltvExpiry,
+                                preimage: ByteVector32,
+                                customTlvs: Seq[GenericTlv] = Nil) extends Recipient {
+  override val totalAmount = amount
+  override val features = Features.empty
+  override val extraEdges = Nil
+
+  override def buildPayloads(paymentHash: ByteVector32, route: Route): Try[PaymentPayloads] = {
+    Recipient.verifyChannelHopsOnly(route).map(hops => {
+      val finalPayload = NodePayload(nodeId, FinalPayload.Standard.createKeySendPayload(route.amount, amount, expiry, preimage, customTlvs))
+      OutgoingPaymentPacket.buildPayloads(amount, expiry, Seq(finalPayload), hops)
+    })
+  }
+}
+
+/** A payment recipient that hides its real identity using route blinding. */
+case class BlindedRecipient(nodeId: PublicKey,
+                            features: Features[InvoiceFeature],
+                            totalAmount: MilliSatoshi,
+                            expiry: CltvExpiry,
+                            extraEdges: Seq[Invoice.BlindedEdge],
+                            customTlvs: Seq[GenericTlv] = Nil) extends Recipient {
+  override def buildPayloads(paymentHash: ByteVector32, route: Route): Try[PaymentPayloads] = {
+    extractRoute(route).map { case (channelHops, blindedHop) =>
+      val introductionPayload = NodePayload(blindedHop.route.introductionNodeId, OutgoingBlindedPerHopPayload.createIntroductionPayload(blindedHop.route.introductionNode.encryptedPayload, blindedHop.route.introductionNode.blindingEphemeralKey))
+      val intermediateBlindedPayloads = blindedHop.route.subsequentNodes.dropRight(1).map(n => NodePayload(n.blindedPublicKey, OutgoingBlindedPerHopPayload.createIntermediatePayload(n.encryptedPayload)))
+      val finalBlindedPayload = NodePayload(blindedHop.route.blindedNodes.last.blindedPublicKey, OutgoingBlindedPerHopPayload.createFinalPayload(route.amount, totalAmount, expiry, blindedHop.route.blindedNodes.last.encryptedPayload, customTlvs))
```

Comment:
This does not work for a zero-hop route.

Reply:
Fixed in the latest version. I added the e2e on the following PR (#2490) because failure handling was really incomplete, so I figured it made sense to add all the e2e tests once we have everything working.